### PR TITLE
530: className must not be null on Android Q

### LINKED ISSUE #530: ViewNode crash: className must not be null on Android Q
Lockbox crashes a lot, always with the following exception:

```
2019-03-22 22:13:25.412 21339-21339/mozilla.lockbox E/AndroidRuntime: FATAL EXCEPTION: main
    Process: mozilla.lockbox, PID: 21339
    java.lang.IllegalStateException: className must not be null
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpNode(ViewNode+.kt:18)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpChildren(ViewNode+.kt:62)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpNode(ViewNode+.kt:46)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpChildren(ViewNode+.kt:62)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpNode(ViewNode+.kt:46)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpChildren(ViewNode+.kt:62)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpNode(ViewNode+.kt:46)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpChildren(ViewNode+.kt:62)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpNode(ViewNode+.kt:46)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpChildren(ViewNode+.kt:62)
        at mozilla.lockbox.extensions.ViewNode_Kt.dumpNode(ViewNode+.kt:46)
        at mozilla.lockbox.extensions.ViewNode_Kt.dump(ViewNode+.kt:10)
        at mozilla.lockbox.LockboxAutofillService.onFillRequest(LockboxAutofillService.kt:71)
        at android.service.autofill.AutofillService.lambda$-android_service_autofill_AutofillService_15826(AutofillService.java:320)
        at android.service.autofill.-$Lambda$svbjmB3NFhHnuZrn67G14PFSJlY.$m$0(Unknown Source:4)
        at android.service.autofill.-$Lambda$svbjmB3NFhHnuZrn67G14PFSJlY.executeMessage(Unknown Source:0)
        at com.android.internal.os.HandlerCaller$MyHandler.handleMessage(HandlerCaller.java:37)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:180)
        at android.app.ActivityThread.main(ActivityThread.java:6950)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:835)
```

The following STR reproduce the crash in 100% for me:

1. make sure Lockbox is enabled in Android autofill options
2. long press to homescreen to add an app icon to the home screen
3. try to enter an app name in the search field

=> Lockbox crashes

### Device information

* Android device: HTC U11 / Android 8.0
* Lockbox version: revision c1f7d1ab474b5312228500bdf612be6aab5273b4

### PR DESCRIPTION
**Pull request by eliserichards**
Fixes #530

## To Do

- [ ] add “WIP” to the PR title if pushing up but not complete nor ready for review
- [ ] double check the original issue to confirm it is fully satisfied
- [ ] add testing notes and screenshots in PR description to help guide reviewers
- [ ] add unit tests
- [ ] make sure CI builds are passing (e.g.: fix lint and other errors)

### CODE REVIEW

**Comment by devinreams** [comment]
Is the DumpNode exception at #597 also caught here?

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 1]
```diff
@@ -68,8 +69,12 @@ class LockboxAutofillService(
...
         if (parsedStructure.passwordId == null && parsedStructure.usernameId == null) {
             if (isDebug()) {
-                val xml = structure.getWindowNodeAt(0).rootViewNode.dump()
-                log.debug("Autofilling failed for:\n$xml")
+                try {
+                    val xml = structure.getWindowNodeAt(0).rootViewNode.dump()
+                    log.debug("Autofilling failed for:\n$xml")
+                } catch (e: Exception) {
+                    log.debug("Exception: $e")
+                }
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: I'm confused. Why is this bug not fixed in `ViewNode.dump()`?

**Comment by eliserichards** [comment]
@devinreams quite possibly...

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 2]
```diff
@@ -32,6 +32,7 @@ import mozilla.lockbox.support.Optional
 import mozilla.lockbox.support.PublicSuffixSupport
 import mozilla.lockbox.support.asOptional
 import mozilla.lockbox.support.isDebug
+import java.lang.Exception
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: Nit: unused import.

File: app/src/main/java/mozilla/lockbox/extensions/ViewNode+.kt [hunk 3]
```diff
@@ -12,7 +12,11 @@ fun AssistStructure.ViewNode.dump(): String {
 
 @TargetApi(Build.VERSION_CODES.O)
 private fun AssistStructure.ViewNode.dumpNode(sb: StringBuilder = StringBuilder()): StringBuilder {
-    val name = htmlInfo?.tag ?: className.split('.').last()
+    val name = if (className.isNullOrBlank()) {
+        htmlInfo?.tag ?: ""
+    } else {
+        htmlInfo?.tag ?: className.split('.').last()
+    }
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/extensions/ViewNode+.kt]: Nit: use optionals and the Elvis operator, and sensible descriptive default:

```kotlin
val name = htmlInfo?.tag ?: className?.split('.')?.last() ?: "unknown"
```

Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/extensions/ViewNode+.kt]: Perfect

**Pull request merged**